core/src/main/java/org/jruby/RubyArithmeticSequence.java

```diff
@@ -0,0 +1,565 @@
+        return new RubyArithmeticSequence(context.runtime, context.runtime.getArithmeticSequence(), generatedBy, method, args, begin, end, step, excludeEnd);
+    }
+
+    // arith_seq_each
+    @JRubyMethod
+    public IRubyObject each(ThreadContext context, Block block) {
+        IRubyObject c = begin;
+        IRubyObject e = end;
+        IRubyObject s = step;
+        IRubyObject len_1, last;
+
+        if (!block.isGiven()) {
+            return this;
+        }
+
+        if (!(step instanceof RubyComplex) && floatStep(context, c, e, s, excludeEnd.isTrue(), true, block)) {
+            return this;
+        }
+
+        if (end.isNil()) {
+            while (true) {
+                block.yield(context, c);
+                c = ((RubyNumeric)c).op_plus(context, s);
+            }
+            //FIXME : unreachable statement
+            //return this;
+        }
+
+        if (Helpers.rbEqual(context, step, int2fix(context.runtime, 0)).isTrue()) {
+            while (true) {
+                block.yield(context, c);
+            }
+            //FIXME : unreachable statement
+            //return this;
+        }
+
+        len_1 = ((RubyNumeric)((RubyNumeric)e).op_minus(context, c)).idiv(context, s);
+        last = ((RubyNumeric)c).op_plus(context, Numeric.f_mul(context, s, len_1));
+        if (excludeEnd.isTrue() && Helpers.rbEqual(context, last, e).isTrue()) {
+            last = ((RubyNumeric)last).op_minus(context, s);
+        }
+
+        if (Numeric.f_negative_p(context, s)) {
+            while (RubyNumeric.numFuncall(context, c, new GeCallSite(), last).isTrue()) {
+                block.yield(context, c);
+                c = ((RubyNumeric)c).op_plus(context, s);
+            }
+        } else {
+            while (RubyNumeric.numFuncall(context, last, new GeCallSite(), c).isTrue()) {
+                block.yield(context, c);
+                c = ((RubyNumeric)c).op_plus(context, s);
+            }
+        }
+
+        return this;
+    }
+
+    // arith_seq_first
+    @JRubyMethod
+    public IRubyObject first(ThreadContext context) {
+        return first(context, null);
+    }
+
+    // arith_seq_first
+    @JRubyMethod
+    public IRubyObject first(ThreadContext context, IRubyObject num) {
+        Ruby runtime = context.runtime;
+        IRubyObject b = begin, e = end, s = step;
+        RubyArray ary;
+        long n;
+        boolean x;
+
+        if (num == null) {
+            if (!e.isNil()) {
+                IRubyObject zero = int2fix(runtime, 0);
+                int r = RubyComparable.cmpint(context, ((RubyNumeric)step).coerceCmp(context, context.sites.Numeric.op_cmp, zero), s, zero);
+                if (r > 0 && Helpers.invokePublic(context, b, ">", e).isTrue()) {
+                    return context.nil;
+                }
+
+                if (r < 0 && Helpers.invokePublic(context, b, "<", e).isTrue()) {
+                    return context.nil;
+                }
```

Comment:
Example of other places where ArithmeticSequence-specific JavaSites sites would be useful.